File: sdk/javascript/src/webhooks.ts
import { createHmac, timingSafeEqual } from "crypto";

export class WebhookVerificationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "WebhookVerificationError";
  }
}

function parseSignatureHeader(header: string): { timestamp: string; signature: string } {
  const parts = new Map(
    header
      .split(",")
      .map((part) => part.split("=") as [string, string])
      .filter(([key, value]) => key && value),
  );
  const timestamp = parts.get("t");
  const signature = parts.get("v1");
  if (!timestamp || !signature) {
    throw new WebhookVerificationError("malformed signature header");
  }
  return { timestamp, signature };
}

export interface VerifyWebhookOptions {
  toleranceSeconds?: number;
}

/**
 * Verifies a SorobanPulse webhook request body against its
 * `X-SorobanPulse-Signature` header (`t=<unix_ts>,v1=<hex_hmac>`), using
 * HMAC-SHA256 over `{timestamp}.{rawBody}`.
 */
export function verifyWebhookSignature(
  rawBody: string | Buffer,
  signatureHeader: string,
  secret: string,
  options: VerifyWebhookOptions = {},
): boolean {
  const toleranceSeconds = options.toleranceSeconds ?? 300;
  const { timestamp, signature } = parseSignatureHeader(signatureHeader);

  const timestampNum = Number(timestamp);
  if (!Number.isFinite(timestampNum)) {
    throw new WebhookVerificationError("invalid timestamp");
  }

  const nowSeconds = Math.floor(Date.now() / 1000);
  if (Math.abs(nowSeconds - timestampNum) > toleranceSeconds) {
    throw new WebhookVerificationError("timestamp outside tolerance window");
  }

  const bodyBuffer = typeof rawBody === "string" ? Buffer.from(rawBody) : rawBody;
  const signedPayload = Buffer.concat([Buffer.from(`${timestamp}.`), bodyBuffer]);
  const expected = createHmac("sha256", secret).update(signedPayload).digest("hex");

  const expectedBuffer = Buffer.from(expected, "hex");
  const signatureBuffer = Buffer.from(signature, "hex");

  if (
    expectedBuffer.length !== signatureBuffer.length ||
    !timingSafeEqual(expectedBuffer, signatureBuffer)
  ) {
    throw new WebhookVerificationError("signature mismatch");
  }

  return true;
}

File: sdk/python/soroban_pulse/webhooks.py
"""Webhook signature verification utilities.

SorobanPulse signs outgoing webhook payloads with HMAC-SHA256 over
`{timestamp}.{body}`, sent as the `X-SorobanPulse-Signature` header in the
form `t=<unix_ts>,v1=<hex_hmac>`.
"""

from __future__ import annotations

import hashlib
import hmac
import time


class WebhookVerificationError(Exception):
    """Raised when a webhook signature fails verification."""


def _parse_signature_header(header: str) -> tuple:
    parts = dict(item.split("=", 1) for item in header.split(",") if "=" in item)
    timestamp = parts.get("t")
    signature = parts.get("v1")
    if not timestamp or not signature:
        raise WebhookVerificationError("malformed signature header")
    return timestamp, signature


def verify_webhook_signature(
    payload: bytes,
    signature_header: str,
    secret: str,
    tolerance_seconds: int = 300,
) -> bool:
    """Verify a webhook request body against its signature header.

    Raises `WebhookVerificationError` if the signature is invalid or the
    timestamp is outside `tolerance_seconds` of the current time (replay
    protection). Returns True on success.
    """
    timestamp, signature = _parse_signature_header(signature_header)

    try:
        ts_int = int(timestamp)
    except ValueError as exc:
        raise WebhookVerificationError("invalid timestamp") from exc

    if abs(time.time() - ts_int) > tolerance_seconds:
        raise WebhookVerificationError("timestamp outside tolerance window")

    signed_payload = f"{timestamp}.".encode("utf-8") + payload
    expected = hmac.new(secret.encode("utf-8"), signed_payload, hashlib.sha256).hexdigest()

    if not hmac.compare_digest(expected, signature):
        raise WebhookVerificationError("signature mismatch")

    return True
